Source/WebSocket/Win32/win32_websocket.cpp

```diff
@@ -3,9 +3,601 @@
+        {
+            ScheduleAsync(async, 0);
+        }
+    }
+
+    // Wrappers for the different types of websocketpp clients.
+    // Perform type erasure to set the websocketpp client in use at runtime
+    // after construction based on the URI.
+    struct websocketpp_client_base
+    {
+        virtual ~websocketpp_client_base() HC_NOEXCEPT {}
+        template <typename WebsocketConfig>
+        websocketpp::client<WebsocketConfig> & client()
+        {
+            if (is_tls_client())
+            {
+                return reinterpret_cast<websocketpp::client<WebsocketConfig> &>(tls_client());
+            }
+            else
+            {
+                return reinterpret_cast<websocketpp::client<WebsocketConfig> &>(non_tls_client());
+            }
+        }
+        virtual websocketpp::client<websocketpp::config::asio_client> & non_tls_client()
+        {
+            throw std::bad_cast();
+        }
+        virtual websocketpp::client<websocketpp::config::asio_tls_client> & tls_client()
+        {
+            throw std::bad_cast();
+        }
+        virtual bool is_tls_client() const = 0;
+    };
+    struct websocketpp_client : websocketpp_client_base
+    {
+        websocketpp::client<websocketpp::config::asio_client> & non_tls_client() override
+        {
+            return m_client;
+        }
+        bool is_tls_client() const override { return false; }
+        websocketpp::client<websocketpp::config::asio_client> m_client;
+    };
+    struct websocketpp_tls_client : websocketpp_client_base
+    {
+        websocketpp::client<websocketpp::config::asio_tls_client> & tls_client() override
+        {
+            return m_client;
+        }
+        bool is_tls_client() const override { return true; }
+        websocketpp::client<websocketpp::config::asio_tls_client> m_client;
+    };
+
+    // Asio client has a long running "run" task that we need to provide a thread for
+    AsyncBlock* m_backgroundAsync;
+
+    websocketpp::connection_hdl m_con;
+
+    websocketpp::lib::error_code m_connectError;
+
+    // Used to safe guard the wspp client.
+    std::mutex m_wsppClientLock;
+    std::atomic<State> m_state;
+    std::unique_ptr<websocketpp_client_base> m_client;
+
+    // Guards access to m_outgoing_msg_queue
+    std::mutex m_outgoingMessageQueueLock;
+
+    // Queue to order the sends
+    http_internal_queue<websocket_outgoing_message> m_outgoingMessageQueue;
+
+    // Number of sends in progress and queued up.
+    std::atomic<int> m_numSends;
+
+    // Used to track if any of the OpenSSL server certificate verifications
+    // failed. This can safely be tracked at the client level since connections
+    // only happen once for each client.
+#if defined(__APPLE__) || (defined(ANDROID) || defined(__ANDROID__)) || defined(_WIN32)
+    bool m_opensslFailed;
+#endif
+
+    hc_websocket_handle_t m_hcWebsocketHandle;
+
+    Uri m_uri;
+
+    const static http_internal_string m_subProtocolHeader;
```

Comment:
don't make static internal strings or it'll break cause static is created before mem hook is set and then its dtor'd using the incorrect free.